### ISSUE TITLE
linkerd-destination: crazy RPS on io.linkerd.proxy.destination.Destination/GetProfile 

### ISSUE THREAD

**Issue by vasi-grigo** [linkerd/linkerd2#5843]
### What is the issue?
- linkerd-destination serving insane number of requests for */io.linkerd.proxy.destination.Destination/GetProfile* that don't seem to be successful.
- linkerd-proxy injected into pod exploding with warnings

### How can it be reproduced?
Has not yet been established.

### Logs, error output, etc
linkerd-proxy:
```console
[  1498.404169s]  WARN ThreadId(01) outbound: linkerd2_service_profiles::client: Could not fetch profile error=status: Unknown, message: "http2 error: protocol error: unexpected internal error encountered", details: [], metadata: MetadataMap { headers: {} }
[  1498.429222s]  WARN ThreadId(01) outbound: linkerd2_service_profiles::client: Could not fetch profile error=status: FailedPrecondition, message: "IP address conflict: &Pod{ObjectMeta:{prometheus-to-sd-5kl8m prometheus-to-sd- kube-system /api/v1/namespaces/kube-system/pods/prometheus-to-sd-5kl8m 78615961-6b97-45dd-84c3-ae7ef3bcb1e3 374477016 0 2021-02-28 16:32:34 +0000 UTC <nil> <nil> map[controller-revision-hash:7c676f4db4 k8s-app:prometheus-to-sd pod-template-generation:10] map[] [{apps/v1 DaemonSet prometheus-to-sd e16c995b-6444-11e9-ae6f-42010a009008 0xc000dcde90 0xc000dcde91}] []  []},
...
```

#### `linkerd check` output

```text
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
√ issuer cert is valid for at least 60 days
√ issuer cert is issued by the trust anchor

linkerd-webhooks-and-apisvc-tls
-------------------------------
√ tap API server has valid cert
√ tap API server cert is valid for at least 60 days
√ proxy-injector webhook has valid cert
√ proxy-injector cert is valid for at least 60 days
√ sp-validator webhook has valid cert
√ sp-validator cert is valid for at least 60 days

linkerd-api
-----------
√ control plane pods are ready
√ control plane self-check
√ [kubernetes] control plane can talk to Kubernetes
√ [prometheus] control plane can talk to Prometheus
√ tap api service is running

linkerd-version
---------------
√ can determine the latest version
√ cli is up-to-date

control-plane-version
---------------------
√ control plane is up-to-date
√ control plane and cli versions match

linkerd-ha-checks
-----------------
√ pod injection disabled on kube-system
√ multiple replicas of control plane pods

linkerd-prometheus
------------------
√ prometheus add-on service account exists
√ prometheus add-on config map exists
√ prometheus pod is running

linkerd-grafana
---------------
√ grafana add-on service account exists
√ grafana add-on config map exists
√ grafana pod is running

Status check results are √
```

### Environment

- Kubernetes Version: 1.16.15-gke.7800
- Cluster Environment: 1.16.15-gke.7800
- Host OS:
- Linkerd version: v2.9.4

### Additional context
<img width="1327" alt="Screenshot 2021-03-01 at 11 49 41" src="https://user-images.githubusercontent.com/4158474/109475694-0fd93980-7a87-11eb-9984-6acb533e35e4.png">
<img width="1396" alt="Screenshot 2021-03-01 at 11 51 50" src="https://user-images.githubusercontent.com/4158474/109475714-15368400-7a87-11eb-9fcc-ad4540bdf8f3.png">

**Comment by vasi-grigo** [comment]
rolled back to 2.8.1 and no longer observing this behavior; nothing else has been changed in the system
so it's somewhere between 2.8.1 and 2.9.4 :)

**Comment by kleimkuhler** [comment]
@vasi-grigo Is your cluster in a state where two pods have the same IP address? This may occur when one pod is terminating and a new one is started—both with the same IP. Ideally this should have been fixed by #5412 which was included in `stable-2.9.2`.

It may be helpful to get pods with the `-o wide` flag cluster wide and see if there are any pods that share the same IP address. If they are both in a running state it may be a different issue to solve.

**Comment by vasi-grigo** [comment]
@kleimkuhler it's somewhat hard to establish that; i'll see if i can reproduce it in our dev cluster
but at the same time, there were 15+ pods that i restarted a couple of times and most were throwing those warnings
so it is kind of unlikely that all of them had an IP address conflict
They were in a running state though and did handle requests.

**Comment by Wenliang-CHEN** [comment]
Hey guys, I think we have the similar issue.  In our case, it seems the traffic is from `linkerd-prometheus.linkerd.serviceaccount.identity.linkerd.cluster.local` to `linkerd-destination`.

<img width="1722" alt="Screenshot 2021-03-02 at 15 56 55" src="https://user-images.githubusercontent.com/17543769/109782422-2909f380-7c09-11eb-8dda-992b877fb53e.png">

We are not able to verify if there are pods with duplicated IPs when this happened.  The amount of request went back to normal after a few restarts of linkerd-prometheus pod.

When describe the linkerd-prometheus pod, we found that the last state of the linkerd-proxy container is OOM killed:

```
      State:         Running
      Started:     Tue, 02 Mar 2021 10:09:53 +0100
    Last State:    Terminated
      Reason:      OOMKilled
      Message:      error encountered", details: [], metadata: MetadataMap { headers: {} }
[425757.102434s]  WARN ThreadId(01) linkerd_service_profiles::client: Could not fetch profile error=status: Unknown, message: "http2 error: protocol error: unexpected internal error encountered", details: [], metadata: MetadataMap { headers: {} }
[425757.104630s]  WARN ThreadId(01) linkerd_service_profiles::client: Could not fetch profile error=status: Unknown, message: "http2 error: protocol error: unexpected internal error encountered", details: [], metadata: MetadataMap { headers: {} }
[425757.104811s]  WARN ThreadId(01) linkerd_service_profiles::client: Could not fetch profile error=status: Unknown, message: "http2 error: protocol error: unexpected internal error encountered", details: [], metadata: MetadataMap { headers: {} }
[425757.106951s]  WARN ThreadId(01) linkerd_service_profiles::client: Could not fetch profile error=status: Unknown, message: "http2 error: protocol error: unexpected internal error encountered", details: [], metadata: MetadataMap { headers: {} }
[425757.107847s]  WARN ThreadId(01) linkerd_service_profiles::client: Could not fetch profile error=status: Unknown, message: "http2 error: protocol error: unexpected internal error encountered", details: [], metadata: MetadataMap { headers: {} }
[425757.108969s]  WARN ThreadId(01) linkerd_service_profiles::client: Could not fetch profile error=status: Unknown, message: "http2 error: protocol error: unexpected internal error encountered", details: [], metadata: MetadataMap { headers: {} }
[425757.110781s]  WARN ThreadId(01) linkerd_service_profiles::client: Could not fetch profile error=status: Unknown, message: "http2 error: protocol error: unexpected internal error encountered", details: [], metadata: MetadataMap { headers: {} }
[425757.114878s]  WARN ThreadId(01) linkerd_service_profiles::client: Could not fetch profile error=status: Unknown, message: "http2 error: protocol error: unexpected internal error encountered", details: [], metadata: MetadataMap { headers: {} }

      Exit Code:    137
      Started:      Thu, 25 Feb 2021 11:53:55 +0100
      Finished:     Tue, 02 Mar 2021 10:09:53 +0100
```

**Comment by kleimkuhler** [comment]
@Wenliang-CHEN Are you also on `stable-2.9.4`? Are you aware of any other events happening on your cluster at that time that may be helpful in figuring out what happened during that time?

Also, has this been a one-time thing or are you consistently seeing this? It'd be helpful to know if you're seeing this on the recent edge releases, so any sort of repro would be helpful.

**Comment by Wenliang-CHEN** [comment]
@kleimkuhler We are on a customized image, which is: stable-2.9.2 + cherry-picked #5722 

During the incident, the communication between the services in the mesh was normal.  We noticed that the linkerd-proxy from linkerd-prometheus emitted huge amount of this type of log: more than 40,000 / min
```
WARN ThreadId(01) linkerd_service_profiles::client: Could not fetch profile error=status: Unknown, message: "http2 error: protocol error: unexpected internal error encountered", details: [], metadata: MetadataMap { headers: {} }
```

We observed 2 incidents of this type in recent 2 weeks.  Both were self-healing after a few hours via pod restart.

For the root cause or trigger, so far we can exclude certificate rotation as the time does not match.  

I will update this thread once we find out more.

**Comment by kleimkuhler** [comment]
@Wenliang-CHEN Okay if you are able to, it may be worth upgrading to `stable-2.9.4`. That release has #5722 as well as some proxy changes that you may not have in 2.9.2.

Once you're on a non-custom image, we may be able to debug a little more since we'll be running the same linkerd installations.

**Comment by Wenliang-CHEN** [comment]
@kleimkuhler we updated to 2.9.4 but the topology issue described in #5684 came back.

And from [the diff](https://github.com/linkerd/linkerd2/compare/stable-2.9.2...stable-2.9.4), it seems #5722 (the fix for #5684) is not included in 2.9.4.  So we rolled back.

In fact you can just take our control plane as 2.9.2.

**Comment by Wenliang-CHEN** [comment]
Hey guys, just a quick update.  In our case, it seems a Prometheus configuration issue.

As described in the previous comment, the crazy amount of request comes from Prometheus.  And in our cluster, we have pods both in and outisde of the mesh.  When then incident happened, we found out from linkerd-proxy log that the traffic was from Prometheus to port 4191 of an IP that is assigned to a pod outside of the mesh.  Once we restarted Prometheus, the load was gone.

The kubernetes_sd_config we had for Prometheus is like this:
```
      kubernetes_sd_configs:
      - role: pod
```

With that setting, Prometheus will try to scrape all the pods.  I suspect it somehow caches the IP of those that return valid data at the first try.  Therefore when the IP gets reassigned, the issue happens.

Then we changed the configuration to the following:
```
      kubernetes_sd_configs:
      - role: pod
        selectors:
        - role: pod
          label: "linkerd.io/control-plane-ns=linkerd"
```

We expect then it only scrapes the pods that are in the mesh.  The issue is gone since we rolled out Prometheus with the new config.

I think then this might not be an issue of Linkerd control plane.

**Comment by kleimkuhler** [comment]
@vasi-grigo This is likely fixed by #5904. I'll update this issue when that is in an edge release this week, but just wanted to let you know we still have an eye on this issue and hope to have a fix in soon.

**Comment by olix0r** [comment]
@vasi-grigo are you able to try there latest edge release? We believe that we've fixed this class of issue. Can you confirm that this is the case? We plan to include this fix in an upcoming stable-2.10.1, so we want to be sure we've actually resolved the issue.

**Comment by kleimkuhler** [comment]
Please reopen if you run into this issue again, but this should be fixed by #5904 which is available in the recent edges and will be in stable 10.1.
